### PR TITLE
Spatially partition projectiles

### LINKED ISSUE #9443: Spatially partition IEffects
We render every effect in the world, including those that are outside the screen bounds.  Adding a new `SpatiallyPartitioned<IEffect> partitionedEffects` to ScreenMap for use by the WorldRenderer should give us a reasonable rendering performance win.  As an added bonus this should cut the rendering workload to the point where it is feasible to interleave actor and effect renderables, which fixes a big chunk of polish issues.

### LINKED ISSUE #14809: Sprite projectiles aren't added to/removed from ScreenMap.
Checked on Bullet, Missile and GravityBomb at least, neither gets the screenmap updates so they don't end up benefitting from the don't-render-outside-of-screen changes.

### PR DESCRIPTION
**Pull request by reaperrr**
This PR finally adresses the last puzzle piece in improving mid-battle render performance by adding all projectiles that are very common and/or support some engine-drawn effect (contrails, beams) to the ScreenMap, meaning they are no longer rendered when off-screen.

`GravityBomb` is now partitioned as well, but mostly because it was simple to add.

Special cases that wouldn't yield worthwhile performance gains (due to being sprite-only) and/or seem too complicated to add have been skipped for now (f.ex. TD `IonCannon`, `TeslaZap`), but in my opinion this should be enough to close #9443.

Fixes #14809.

### CODE REVIEW

**Comment by reaperrr** [comment]
My own impression is that some of my bounds caclulations are likely a bit too complicated plus a bit overkill in terms of safety margins, but I couldn't come up with anything better. Feel free to make suggestions on how to improve/simplify them.

File: OpenRA.Mods.Common/Projectiles/Missile.cs [hunk 1]
```diff
@@ -251,10 +255,23 @@ public Missile(MissileInfo info, ProjectileArgs args)
 
 			if (info.ContrailLength > 0)
 			{
+				hasContrail = true;
+
+				// Approximate maximum contrail length in WDist, used to calculate ScreenMap bounds.
+				// The *3 /2 magic is a HACK to work around the fact that speed * ContrailLength does not actually result in wide enough bounds.
+				// TODO: Figure out a way to calculate this more accurately and less hacky.
+				maxVisualContrailLength = new WDist((info.ContrailLength + info.ContrailDelay) * maxSpeed * 3 / 2);
```
Comment (Unit158) [inline OpenRA.Mods.Common/Projectiles/Missile.cs]: I might look into this, if you haven't tried solving it yet

Comment (reaperrr) [reply OpenRA.Mods.Common/Projectiles/Missile.cs]: Feel free to, I probably won't have time before Sunday and have no idea how to tackle this anyway.

File: OpenRA.Mods.Common/Projectiles/AreaBeam.cs [hunk 2]
```diff
@@ -148,6 +151,11 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
 			target += dir * info.BeyondTargetRange.Length / 1024;
 
 			length = Math.Max((target - headPos).Length / speed.Length, 1);
+
+			beamCenter = WPos.Lerp(headPos, tailPos, 1, 2);
+			beamDelta = headPos - tailPos;
+			screenMapBounds = Util.ProjectileScreenBounds(world, null, info.Width, beamDelta);
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Projectiles/AreaBeam.cs]: Seem like you could extract a method for the above.

File: OpenRA.Mods.Common/Projectiles/AreaBeam.cs [hunk 3]
```diff
@@ -93,13 +93,16 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo wi)
...
+	public class AreaBeam : IProjectile, ISync, ISpatiallyPartitionable
 	{
 		readonly AreaBeamInfo info;
 		readonly ProjectileArgs args;
 		readonly AttackBase actorAttackBase;
 		readonly Color color;
 		readonly WDist speed;
+		Size screenMapBounds;
+		WPos beamCenter;
+		WVec beamDelta;
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Projectiles/AreaBeam.cs]: Do these need to be fields? It looks like you don't actually need to maintain them beyond passing them directly to the screen map - would they be fine as locals?

File: OpenRA.Mods.Common/Projectiles/Bullet.cs [hunk 4]
```diff
@@ -174,10 +178,23 @@ public Bullet(BulletInfo info, ProjectileArgs args)
 
 			if (info.ContrailLength > 0)
 			{
+				hasContrail = true;
+
+				// Approximate maximum contrail length in WDist, used to calculate ScreenMap bounds.
```
Comment (pchote) [inline OpenRA.Mods.Common/Projectiles/Bullet.cs]: Wouldn't it be better to update the `ContrailRenderable` to return a proper bounding rect, and then intersect that with the sprite bounds?

Comment (reaperrr) [reply OpenRA.Mods.Common/Projectiles/Bullet.cs]: Probably, but judging by the compile errors I get after 2 hours of working on this, it looks like I have no clue how to do that.

**Comment by reaperrr** [comment]
Updated, fixed everything except the contrail bounds calculations (don't know how to translate that request into working code).

**Comment by pchote** [comment]
Can we split the contrail stuff off to a different PR so that we can merge the rest more easily?

**Comment by reaperrr** [comment]
> Can we drop or split the contrail stuff off to a different PR so that we can merge the rest more easily?

Sure, though we'd have to either 
a) temporarily live with bullet/missile contrails disappearing as soon as the projectile sprite is off-screen, or
b) drop partitioning from those 2 projectiles completely until we have proper contrail partitioning (which would considerably reduce the performance gain from this PR, as those are the 2 most common projectiles)

**Comment by reaperrr** [comment]
Updated.

**Comment by pchote** [comment]
I hope you don't mind (and suspect that you'll agree) that we still want this in the medium term, but probably can't justify delaying the playtest (which now has many high priority fixes) for it.  Bumping to Next + 1, and I can give some help if you need it once the release pressure goes down.

**Comment by reaperrr** [comment]
Yeah, I'm perfectly fine with that and was thinking about suggesting just that myself.

**Comment by reaperrr** [comment]
I have too many other things I consider more urgent on my near-term roadmap, closing for now.

**Pull request closed without merging**